### PR TITLE
Fetch FG Goods based on Material Transfer

### LINKED ISSUE #3340: Stock in BOM is fetched when updating finished goods, not from stock entries
Actual Stock Issued is not Carried over during Update finished Goods
Stock in BOM is fetched when updating finished goods, not from stock entries

The scenario is stock is issued at different times for production order
not all stock is issued at once 
so I made 2 different stock entries on same production order and noticed when i click update finished goods, the items were fetched from BOM , not from stock entries to that production order

In BOM of Table
GLUE-ICI is 2 Kg

but during stock issuance, it was actually used 2.5 kg, **in 2 different stock enteries**
Both linked to same production order

**when i press update finished goods, it only shows 2kg as per BOM, not 2.5 Kg which is actually issued**
see pic

**BOM Table**
-------------

https://www.dropbox.com/s/9jitjpr6x3fw7v6/bom%20table.jpg?dl=0

**Stock Entry 1**
-----------------

https://www.dropbox.com/s/yit36tffsoqinjn/table%20Stock%20Entry%201.jpg?dl=0

**Stock Entry 2.**
------------------

https://www.dropbox.com/s/1ic33wdn3d3b2x5/table%20Stock%20Entry%202.jpg?dl=0

**Finished Item**
-----------------

https://www.dropbox.com/s/7n57aihq9dspnpn/Table%20Finished.jpg?dl=0

### PR DESCRIPTION
**Pull request by neilLasrado**
Fixed #3340

### CODE REVIEW

**Comment by anandpdoshi** [comment]
@nabinhait please review this code

File: erpnext/stock/doctype/stock_entry/stock_entry.py [hunk 1]
```diff
@@ -643,8 +647,52 @@ def get_bom_raw_materials(self, qty):
...
+		for item in items_dict:
+			qty= item.qty
+			
+			if manufacturing_qty > (produced_qty + flt(self.fg_completed_qty)):
+				qty = (qty/trans_qty) * flt(self.fg_completed_qty)
+			
+			elif transfered_qty.get(item.item_code):
+				qty-= transfered_qty.get(item.item_code)
+				
+			if qty > 0:
```
Comment (anandpdoshi) [inline erpnext/stock/doctype/stock_entry/stock_entry.py]: Just a thought: is there any possibility of this qty going negative?

File: erpnext/manufacturing/doctype/manufacturing_settings/manufacturing_settings.json [hunk 2]
```diff
@@ -34,6 +34,15 @@
    "precision": ""
   }, 
   {
+   "default": "BOM", 
+   "fieldname": "update_fg_based_on", 
+   "fieldtype": "Select", 
+   "label": "Update FG Based On",
```
Comment (nabinhait) [inline erpnext/manufacturing/doctype/manufacturing_settings/manufacturing_settings.json]: Backflush Raw Materials Based On (better label?)

- BOM
- Material Transferred for Manufacture

@rmehta please suggest

**Comment by cgpurbaugh** [comment]
Will this pull batch numbers from the stock entries?

File: erpnext/stock/doctype/stock_entry/stock_entry.py [hunk 3]
```diff
@@ -643,8 +647,52 @@ def get_bom_raw_materials(self, qty):
 
 		for item in item_dict.values():
 			item.from_warehouse = self.from_warehouse or item.default_warehouse
-
 		return item_dict
+		
+	def get_transfered_raw_materials(self):
+		items_dict = frappe.db.sql("""select item_name, item_code, sum(qty) as qty, to_warehouse, from_warehouse,
```
Comment (nabinhait) [inline erpnext/stock/doctype/stock_entry/stock_entry.py]: Use s_warehouse and t_warehouse from item table

File: erpnext/stock/doctype/stock_entry/stock_entry.py [hunk 4]
```diff
@@ -643,8 +647,52 @@ def get_bom_raw_materials(self, qty):
...
+			description, stock_uom, expense_account, cost_center from `tabStock Entry` se,`tabStock Entry Detail` sed 
+			where se.name = sed.parent and se.docstatus=1 and se.purpose='Material Transfer for Manufacture' and 
+			se.production_order= %s and se.to_warehouse= %s group by sed.item_code""", 
+			(self.production_order, self.from_warehouse), as_dict=1)
+		
+		transfered_materials = frappe.db.sql("""select item_code, sum(qty) as qty from `tabStock Entry` se,
+			`tabStock Entry Detail` sed where se.name = sed.parent and se.docstatus=1 and 
+			se.purpose='Manufacture' and se.production_order= %s and se.from_warehouse= %s
```
Comment (nabinhait) [inline erpnext/stock/doctype/stock_entry/stock_entry.py]: check only ifnull(sed.s_warehouse, '') != ''

**Pull request closed without merging**